The goal of this script is to use the JTK_Cycle to discover the rhythmic transcripts in my data, and to analyze if there is any difference between constant and fluctuating temperature.


# 0. Loading packages
```{r}
library(edgeR)

```


# 1. PreProcessing
### Import the data and creat a DGEList object:
```{r}
counts <- read.delim("C:/Users/Po-Kai/Dropbox/Working/Mimulus_project/RNAseq/all_counts.txt", row.names = 1)

d0 <- DGEList(counts)


# Add sample-level info:
group <- as.factor(c(paste("c",rep(1:9, each=4), sep=""),paste("f",rep(1:9, each=4), sep="")))
d0$samples$group <- group

#ZT_vector <- c(14,17,20,23,2,5,8,11,14)
ZT_vector <- c(-10,-7,-4,-1,2,5,8,11,14)
ZT <- rep(rep(ZT_vector, each=4),2)
d0$samples$ZT <- ZT

trt <- as.factor(c(rep(c("c","f"),each=36)))
d0$samples$trt <- trt

```


### Filter the data:
Use the `filterByExpr` function in the `edgeR` package to do the filtering. (20190419)
```{r}
keep.exprs <- filterByExpr(d0,group=d0$samples$group)
#keep.exprs <- filterByExpr(d0)
# These two are the same if I've set the `group=d0$samples$group`.

d0_filterByExpr <- d0[keep.exprs, , keep.lib.sizes=FALSE]
dim(d0_filterByExpr)
```


### Normalized the data and caculate the cpm
```{r}
d0_filterByExpr <- calcNormFactors(d0_filterByExpr)

cpm_filtered <- as.data.frame(cpm(d0_filterByExpr))
```


### Re-organize the data for the JTK-Cycle
```{r}
# Separate the data of constant and fluctuating trt:
data_c <- cpm_filtered[,1:36]
data_f <- cpm_filtered[,37:72]

# Rename the colnames:
#cn <- c(paste(rep(paste("ZT", seq(14,by=3,length.out=9), sep=""), each=4),paste("rep", rep(1:4), sep=""), sep="_"))
cn <- c(paste(rep(paste("ZT", seq(-10,by=3,length.out=9), sep=""), each=4),paste("rep", rep(1:4), sep=""), sep="_"))
colnames(data_c) <- cn
colnames(data_f) <- cn

```



# 2. Run the JTK_Cycle
This part is modified from ¡§Run_JTK_CYCLE (Example2).R¡¨ provided by JTK_Cycle authors.


### Set up some variables:
Set up the working directory for sourcing a script if needed. (Line 1~8 of the original script.)

```{r}
source("JTK_CYCLEv3.1.R")

#project <- "SWC_JTK_Cycle_c"  # I put this in the chunk of "Run the analysis" below.

# This is a setting for preventing some data.frame loading issues.
options(stringsAsFactors=FALSE)

# Set up the `annot` and `data` which are written in the script we sourced.
annot <- as.data.frame(rownames(data_c))  # The annot is the same for both c and f conditions.
colnames(annot) <- "genes"
#data <- data_c # I put this in the chunk of "Run the analysis" below.

```


### Set the parameters for the analysis:
```{r}
# 9 total time points, 4 replicates per time point
jtkdist(9, 4)


# Looking for rhythms of 24 hours. (We used the LD conditions so we are only interested in 24h-rhythms.)
periods <- 8       


# 3 is the number of hours between time points
jtk.init(periods,3) 

```


### Run the analysis: 
1. Run the constant tmp first:
```{r}
# Set up the `project` and `data`
project <- "SWC_JTK_Cycle_c"
data <- data_c

# Run
cat("JTK analysis started on",date(),"\n")
flush.console()

st <- system.time({
  res <- apply(data,1,function(z) {
    jtkx(z)
    c(JTK.ADJP,JTK.PERIOD,JTK.LAG,JTK.AMP)
  })
  res <- as.data.frame(t(res))
  bhq <- p.adjust(unlist(res[,1]),"BH")
  res <- cbind(bhq,res)
  colnames(res) <- c("BH.Q","ADJ.P","PER","LAG","AMP")
  results_c <- cbind(annot,res,data)
  results_c <- results_c[order(res$ADJ.P,-res$AMP),]
})
print(st)


# Export the results:
#save(results_c,file=paste("JTK",project,"rda",sep="."))
#write.table(results_c,file=paste("JTK",project,"txt",sep="."),row.names=F,col.names=T,quote=F,sep="\t")

```
- Observations: 21,731
- Variables: 42 (36+6)

After checking the results with `tail(results_c)`, we know that we have to filter the results by BH.Q or ADJ.P!


2. Run the fluctuating tmp:
```{r}
# Set up the `project` and `data`
project <- "SWC_JTK_Cycle_f"
data <- data_f

# Run
cat("JTK analysis started on",date(),"\n")
flush.console()

st <- system.time({
  res <- apply(data,1,function(z) {
    jtkx(z)
    c(JTK.ADJP,JTK.PERIOD,JTK.LAG,JTK.AMP)
  })
  res <- as.data.frame(t(res))
  bhq <- p.adjust(unlist(res[,1]),"BH")
  res <- cbind(bhq,res)
  colnames(res) <- c("BH.Q","ADJ.P","PER","LAG","AMP")
  results_f <- cbind(annot,res,data)
  results_f <- results_f[order(res$ADJ.P,-res$AMP),]
})
print(st)


# Export the results:
#save(results_f,file=paste("JTK",project,"rda",sep="."))
#write.table(results_f,file=paste("JTK",project,"txt",sep="."),row.names=F,col.names=T,quote=F,sep="\t")

```


3. Filter the data:
```{r}
# Filter for the sig traits 
## Modified Dan's code:
#sig_traits = function(results,p_value_cutoff){
#  return(rownames(results)[results[,"BH.Q"] < p_value_cutoff])
#}

p_value_cutoff <- 1e-3
results_c_filtered <- subset(results_c, BH.Q < p_value_cutoff)
results_f_filtered <- subset(results_f, BH.Q < p_value_cutoff)

```
1. Using 1e-3 as the cutoff to filter my data, I got 1478 and 2795 genes under c and f conditions, respectively.
> Update (20190423, using filterByExpr) I got 1557 and 2945 genes under c and f conditions, respectively.


# 3. Heatmap:
Using pheatmap:
```{r}
library(pheatmap)
```


### Create the data matrix
```{r}
# Sort the data according to their LAG (phase).
results_c_filtered = results_c_filtered[order(results_c_filtered$LAG),]
results_f_filtered = results_f_filtered[order(results_f_filtered$LAG),]


# Matrix:
data_matrix_c <- data.matrix(results_c_filtered[,-c(1:6)])
data_matrix_f <- data.matrix(results_f_filtered[,-c(1:6)])


# Color
colfunc <- colorRampPalette(c("blue", "yellow"))


# Heatmap!
pheatmap(data_matrix_c, scale="row", color=colfunc(15), cluster_rows=FALSE, cluster_cols=FALSE,show_rownames = F )
pheatmap(data_matrix_f, scale="row", color=colfunc(15), cluster_rows=FALSE, cluster_cols=FALSE,show_rownames = F )

```


# 4. Scatter plot:
```{r}
f_scatter <- results_f_filtered[1:6]
f_scatter$group <- c("f")

c_scatter <- results_c_filtered[1:6]
c_scatter$group <- c("c")

full_scatter <- rbind(c_scatter, f_scatter)

cal_z_score <- function(x){
  (x - mean(x)) / sd(x)
}

#data_subset_norm <- t(apply(data_subset, 1, cal_z_score))


# LAG vs. AMP
library(ggplot2)

ggplot(full_scatter, aes(x=LAG, y=AMP, color=group)) +
  geom_point()


# LAG vs. ADJ.P
ggplot(full_scatter, aes(x=LAG, y=ADJ.P, color=group)) +
  geom_point()

```
1. I can not use the AMP without scaling for creating the plot! The variation is too big!

2. My plots didn't provide useful info...


# 5. Discuss with Dan: We would like to know the distribution of the phases
```{r}
unique(results_c_filtered$LAG)
table(results_c_filtered$LAG)


# Visualize the distribution of the phases:
plot(table(results_c_filtered$LAG))
plot(table(results_f_filtered$LAG))


# Check the AMP:
plot(density(results_c_filtered$AMP))
plot(density(log(results_c_filtered$AMP)))
lines(density(log(results_f_filtered$AMP)),col=2)


summary(results_c_filtered$AMP)
summary(results_f_filtered$AMP)
```
1. The distribution of the phases is consistent with the heatmap: there are 2 peaks in f condition, but the c condition only has a peak.

2. The program may fit the 3/2=1.5 into each time point I collected the samples. The numbers of those points are always lower.

3. The AMP is slightly smaller in f condition.

4. Note: The original version Dan and I forgot to use the filtered results.


# 6. Check the heatmap pattern again:

6.1 Checking the heatmap pattern of the 2945 rhythmic genes detected using f samples under the c conditon:
```{r}
f.filtered_c = results_c[rownames(results_c) %in% results_f_filtered$genes,]


# Sort the data according to their LAG (phase).
f.filtered_c = f.filtered_c[order(f.filtered_c$LAG),]


# Matrix:
f.filtered_c.matrix <- data.matrix(f.filtered_c[,-c(1:6)])


# Color
colfunc <- colorRampPalette(c("blue", "yellow"))


# Heatmap!
pheatmap(f.filtered_c.matrix, scale="row", color=colfunc(15), cluster_rows=FALSE, cluster_cols=FALSE,show_rownames = F )


# Bar chart:
plot(table(f.filtered_c$LAG))


```


6.2 Checking the heatmap pattern of the 1557 rhythmic genes detected using c samples under the f conditon:
```{r}
c.filtered_f = results_f[rownames(results_f) %in% results_c_filtered$genes,]


# Sort the data according to their LAG (phase).
c.filtered_f = c.filtered_f[order(c.filtered_f$LAG),]


# Matrix:
c.filtered_f.matrix <- data.matrix(c.filtered_f[,-c(1:6)])


# Color
colfunc <- colorRampPalette(c("blue", "yellow"))


# Heatmap!
pheatmap(c.filtered_f.matrix, scale="row", color=colfunc(15), cluster_rows=FALSE, cluster_cols=FALSE,show_rownames = F )


# Bar chart:
plot(table(c.filtered_f$LAG))


```


# 7. Venn Diagram:

```{r}
library(VennDiagram)
library(grid)

venn.plot <- venn.diagram(list(rownames(results_c_filtered), rownames(results_f_filtered)), NULL, lty=0.1,fill=c("light blue", "pink"), alpha=c(0.5,1), cex = 1.5, cat.fontface=20, category.names=c("Constant tmp", "Fluctuating tmp"), margin = 0.05, cat.pos=c(-30,30), cat.dist=0.05)


grid.draw(venn.plot)
```


